alchemist-web-renderer/src/jvmMain/kotlin/it/unibo/alchemist/server/routes/EnvironmentRoute.kt

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2010-2022, Danilo Pianini and contributors
+ * listed, for each module, in the respective subproject's build.gradle.kts file.
+ *
+ * This file is part of Alchemist, and is distributed under the terms of the
+ * GNU General Public License, with a linking exception,
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
+package it.unibo.alchemist.server.routes
+
+import io.ktor.server.routing.Route
+import io.ktor.server.routing.get
+import it.unibo.alchemist.common.model.serialization.encodeEnvironmentSurrogate
+import it.unibo.alchemist.common.model.serialization.jsonFormat
+import it.unibo.alchemist.common.renderer.Bitmap32Serializer
+import it.unibo.alchemist.common.utility.Ruotes.environmentClientPath
+import it.unibo.alchemist.common.utility.Ruotes.environmentServerPath
+import it.unibo.alchemist.server.state.ServerStore.store
+import it.unibo.alchemist.server.utility.Response
+import it.unibo.alchemist.server.utility.Response.Companion.respond
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+/**
+ * Logic of the Routes in the /environment path.
+ */
+object EnvironmentRoute {
+
+    /**
+     * Route of type GET that retrieve current Environment.
+     * The server will render the environment and send it to the client in an already rendered form.
+     * The HTTP [Response] sent to the client can be of type:
+     * 200 (OK) the Environment is sent to the client.
+     */
+    fun Route.environmentServerMode() {
+        get(environmentServerPath) {
+            respond(Response(content = renderedEnvironment()))
+        }
+    }
+
+    /**
+     * Route of type GET that retrieve current Environment.
+     * The server will send the environment to the client in a serialized form.
+     * The HTTP [Response] sent to the client can be of type:
+     * 200 (OK) the Environment is sent to the client.
+     */
+    fun Route.environmentClientMode() {
+        get(environmentClientPath) {
+            respond(Response(content = jsonFormat.encodeEnvironmentSurrogate(store.state.environmentSurrogate)))
```

Comment:
Some [formats](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/formats.md) are still experimental. Protobuf can be added as another dependency and uses different APIs. There is not a way to strategy-out the serializer used natively, but it certainly can be implemented by hand. However, I would personally wait until those dependencies are not experimental anymore, especially if the intent is to use them in production.

Reply:
I see your point, but the concept of "experimental" in Kotlin is more stable than the concept of "stable" in Alchemist, being the latter a simulator for scientific research :wink: 

So in kotlinx.serialization every serializer exposes a different API?

Reply:
Yes, please check these links with some examples:
[Protobuf API example](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/formats.md#protobuf-experimental)
[CBOR API example](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/formats.md#cbor-experimental)
As you can see the Classes used here differs, same thing happens for Json. There's one thing I didn't mention though, don't know why I didn't say it first. Remember this?
```kotlin
install(ContentNegotiation) {
    json()
}
```
By using this code in the Ktor install module we instruct the routes on the server to automatically use serializers if the object that we want to send is an instance of a serializable class. Same thing happens on the client side for deserialization. In other words, I could have done a simple:
```kotlin
fun Route.environmentClientMode() {
  get(environmentClientPath) {
      respond(Response(store.state.environmentSurrogate))
  }
}
 ```
There's a problem though. The ``EnvironmentSurrogate`` class needs a specific Polymorphic serializer because of the two type parameters. Ktor still does not support Polymorphic serializers in content negotiation, so I had to manually create that annoying ``encodeEnvironmentSurrogate`` extension function.
This was the bad news. 
The good news is: changing that ``json()`` to ``protobuf()`` would be enough to accomplish what you want, but only if this limit is surpassed in some way. For now, the only valid open option that comes to my mind is to create a function that would manually check the Content Negotiation strategy being used (hoping this is possible) and change the serialization API accordingly.

Also note that the API changes, but the structure of the Surrogate Classes would remain the same.